### DISCUSSION TITLE
Automatically Adapt By Name Parameters

### DISCUSSION THREAD

**Discussion by adamgfraser** [scala/scala3#17834] [Feature Requests]
Currently the code below does not compile. `printLine` takes a by name argument whereas `flatMap` expects a function that takes an argument that is not by name. We can resolve the compilation error by replacing `printLine` with `printLine(_)` but it would be nice if the compiler allowed us to do this without the additional noise.

```scala
trait ZIO[-R, +E, +A]:
  def flatMap[R1 <: R, E1 >: E, B](f: A => ZIO[R1, E1, B]): ZIO[R1, E1, B] =
    ???

def printLine(line: => Any): ZIO[Any, Nothing, Unit] =
  ???

val zio1: ZIO[Any, Nothing, Int] =
  ???

val zio2 = zio1.flatMap(printLine)

// Found:    (=> Any) => ZIO[Nothing, Any, Any]
// Required: Int => ZIO[Nothing, Any, Any]
```

**Comment by SethTisue** [comment]
I didn't turn up an identical Scala 2 ticket in scala/bug. but I'll note https://github.com/scala/bug/issues/9313 in passing anyway, since it's related in the sense that it's also about the (IMO) rather murky, semi-supported nature of by-name parameters to function values

**Comment by bbarker** [comment]
Cross-referencing proposal on Discourse: [Make by-name parameters just sugar syntax rather than actual types](https://contributors.scala-lang.org/t/make-by-name-parameters-just-sugar-syntax-rather-than-actual-types/5228)
